### ISSUE TITLE
ArC: support async `PostConstruct`?

### ISSUE THREAD

**Issue by FroMage** [quarkusio/quarkus#2384]
As seen in https://github.com/quarkusio/quarkus/pull/1816/files#diff-31226afb2b118a1ed2d1a0c11980d202R147 some users will have async tasks to do in `@PostConstruct` methods. The workaround is to tell them to do a `join()` on them, effectively blocking startup until the task is completed.

It's not very important I think in most applications to block startup, but we could perhaps support `@PostConstruct` methods that return `CompletionStage` where ArC would take care of waiting for the task to be done. Initially it could do it as simply as calling `toCompletableFuture().join()` on them, that's totally fine. Later we could imagine collecting all such async `@PostConstruct` tasks and letting them run in parallel until they're all done and we can proceed to start the application.

In other words, even if it does not gain us performance (because the container is still blocking) it removes boilerplate from user code.

WDYT?

@mkouba @cescoffier

**Comment by dmlloyd** [comment]
Could `@Priority` cause a problem there?

**Comment by mkouba** [comment]
Hm, so first of all `@PostConstruct` != application startup. Even for `@ApplicationScoped` beans this callback is invoked when the contextual instance is being created which is the first time a method is invoked upon the instance (be it an observer method or a business method invocation).

> @PostConstruct methods that return CompletionStage where ArC would take care of waiting for the task to be done.

This would work but I'm not sure it's worth the effort because `toCompletableFuture().join()` is the only boilerplate we would remove.

> Later we could imagine collecting all such async @PostConstruct tasks...

You can't do that for a single component because interceptors build up a chain that must be processed sequentially (and yes @dmlloyd, `@Priority`must be obeyed).

Anyway, you know my opinion on the "async bean construction stuff" - we would have to rework every piece of the CDI SPI otherwise it does not make much sense. We could experiment with ArC APIs though. The only problem is where to find the spare cycles...

I'm not sure what others think? @cescoffier @manovotn

**Comment by gsmet** [comment]
Closing as it seems we won't get to it.
